main/src/ca/uwaterloo/flix/tools/Tester.scala

```diff
@@ -1,125 +1,208 @@
+    /**
+      * Runs all the given tests.
+      */
+    override def run(): Unit = {
+      val start = System.nanoTime()
+      for (testCase <- tests) {
+        runTest(testCase)
       }
-      sb.toString()
+      val elapsed = System.nanoTime() - start
+      queue.add(TestEvent.Finished(Duration(elapsed)))
     }
 
-    def overallResult: OverallTestResult = {
-      if (results.isEmpty) {
-        OverallTestResult.NoTests
-      } else if (results.forall(_.isInstanceOf[TestResult.Success])) {
-        OverallTestResult.Success
-      } else {
-        OverallTestResult.Failure
-      }
-    }
-  }
+    /**
+      * Runs the given `test` emitting test events.
+      */
+    private def runTest(test: TestCase): Unit = test match {
+      case TestCase(sym, run) =>
+        queue.add(TestEvent.Before(sym))
 
-  /**
-   * Evaluates all tests.
-   *
-   * Returns a pair of (successful, failed)-tests.
-   */
-  def test(compilationResult: CompilationResult): TestResults = {
-    val results = compilationResult.getTests.toList.map {
-      case (sym, defn) =>
+        val start = System.nanoTime()
         try {
-          val result = defn()
+          val result = run()
+          val elapsed = System.nanoTime() - start
+
           result match {
-            case java.lang.Boolean.TRUE => TestResult.Success(sym, "Returned true.")
-            case java.lang.Boolean.FALSE => TestResult.Failure(sym, "Returned false.")
-            case _ => TestResult.Success(sym, "Returned non-boolean value.")
+            case java.lang.Boolean.TRUE =>
+              queue.add(TestEvent.Success(sym, Duration(elapsed)))
+
+            case java.lang.Boolean.FALSE =>
+              queue.add(TestEvent.Failure(sym, Duration(elapsed)))
+
+            case _ =>
+              queue.add(TestEvent.Success(sym, Duration(elapsed)))
+
           }
         } catch {
           case ex: Exception =>
-            TestResult.Failure(sym, ex.getMessage)
+            val elapsed = System.nanoTime() - start
+            queue.add(TestEvent.Failure(sym, Duration(elapsed)))
         }
     }
-    TestResults(results)
+  }
+
+  /**
+    * Returns all test cases from the given compilation `result`.
+    */
+  private def getTestCases(compilationResult: CompilationResult): Vector[TestCase] =
+    compilationResult.getTests.toVector.sortBy(_._1.toString).map {
+      case (sym, defn) => TestCase(sym, defn)
+    }
+
+  /**
+    * Represents a single test case.
+    *
+    * @param sym the Flix symbol.
+    * @param run the code to run.
+    */
+  case class TestCase(sym: Symbol.DefnSym, run: () => AnyRef)
+
+  /**
+    * A common super-type for test events.
+    */
+  sealed trait TestEvent
+
+  object TestEvent {
+
+    /**
+      * A test event emitted immediately before a test case is executed.
+      */
+    case class Before(sym: Symbol.DefnSym) extends TestEvent
```

Comment:
What's this for?

Reply:
To print that we are "working on" the test. But not used yet.